src/components/decision-engine/index.js

```diff
@@ -0,0 +1,265 @@
+const Wantlist = require('../../types/wantlist')
+const Ledger = require('./ledger')
+
+class DecisionEngine {
+  constructor (blockstore, network) {
+    this.blockstore = blockstore
+    this.network = network
+
+    // A list of of ledgers by their partner id
+    this.ledgerMap = new Map()
+    this._running = false
+
+    // List of tasks to be processed
+    this._tasks = []
+
+    this._outbox = debounce(this._processTasks.bind(this), 100)
+  }
+
+  _sendBlocks (env, cb) {
+    const msg = new Message(false)
+
+    env.blocks.forEach((b) => {
+      msg.addBlock(b.cid, b.block)
+    })
+
+    // console.log('sending %s blocks', msg.blocks.size)
+    this.network.sendMessage(env.peer, msg, (err) => {
+      if (err) {
+        log('sendblock error: %s', err.message)
+      }
+      cb()
+    })
+  }
+
+  _processTasks () {
+    if (!this._running || !this._tasks.length) {
+      return
+    }
+
+    const tasks = this._tasks
+    this._tasks = []
+    const entries = tasks.map((t) => t.entry)
+    const cids = entries.map((e) => e.cid)
+    const uniqCids = uniqWith(cids, (a, b) => a.equals(b))
+    const groupedTasks = groupBy(tasks, (task) => task.target.toB58String())
+
+    waterfall([
+      (cb) => map(uniqCids, (cid, cb) => {
+        pull(
+          this.blockstore.getStream(cid.multihash),
+          pull.collect((err, blocks) => {
+            if (err) {
+              return cb(err)
+            }
+            cb(null, {
+              cid: cid,
+              block: blocks[0]
+            })
+          })
+        )
+      }, cb),
+      (blocks, cb) => each(values(groupedTasks), (tasks, cb) => {
+        // all tasks have the same target
+        const peer = tasks[0].target
+        const blockList = cids.map((cid) => {
+          return find(blocks, (b) => b.cid.equals(cid))
+        })
+
+        this._sendBlocks({
+          peer: peer,
+          blocks: blockList
+        }, (err) => {
+          if (err) {
+            log.error('failed to send', err)
+          }
+          blockList.forEach((block) => {
+            this.messageSent(peer, block.block, block.cid)
+          })
+          cb()
+        })
+      })
+    ], (err) => {
+      this._tasks = []
+      if (err) {
+        log.error(err)
+      }
+    })
+  }
+
+  wantlistForPeer (peerId) {
+    const peerIdStr = peerId.toB58String()
+    if (!this.ledgerMap.has(peerIdStr)) {
+      return new Map()
+    }
+
+    return this.ledgerMap.get(peerIdStr).wantlist.sortedEntries()
+  }
+
+  peers () {
+    return Array.from(this.ledgerMap.values()).map((l) => l.partner)
+  }
+
+  receivedBlocks (cids) {
+    if (!cids.length) {
+      return
+    }
+    // Check all connected peers if they want the block we received
+    for (let l of this.ledgerMap.values()) {
+      cids
+        .map((k) => l.wantlistContains(k))
+        .filter(Boolean)
+        .forEach((e) => {
+          // this.peerRequestQueue.push(e, l.partner)
```

Comment:
Which is important to keep on a per peer basis, as we send messages out on a per peer basis.